index.js

```diff
@@ -1,207 +1,185 @@
+  const { options: overrideConfig } = options
+  let content
+  if (type === 'module') {
+    content = await import(file)
+  } else {
+    content = require(file)
+  }
+  const plugin = wrapRoutes(content.default || content)
+  const pluginConfig = (content.default && content.default.autoConfig) || content.autoConfig || {}
+  const pluginOptions = Object.assign({}, pluginConfig, overrideConfig)
+  const pluginMeta = plugin[Symbol.for('plugin-meta')] || {}
 
-        fastify.register(plugin.default || plugin, options)
-        loadedPlugins[name] = true
-      }
+  if (plugin.autoload === false || content.autoload === false) {
+    return
+  }
 
-      let cyclicDependencyCheck = {}
+  // Reset to support overriding autoConfig for library plugins
+  if (plugin.autoConfig !== undefined) {
+    plugin.autoConfig = undefined
+  }
 
-      function loadPlugin ({ plugin, name, dependencies = [], options }) {
-        if (cyclicDependencyCheck[name]) throw new Error('Cyclic dependency')
+  const prefixOverride = plugin.prefixOverride !== undefined ? plugin.prefixOverride : content.prefixOverride !== undefined ? content.prefixOverride : undefined
+  const prefix = (plugin.autoPrefix !== undefined ? plugin.autoPrefix : content.autoPrefix !== undefined ? content.autoPrefix : undefined) || directoryPrefix
+  if (prefixOverride !== undefined) {
+    pluginOptions.prefix = prefixOverride
+  } else if (prefix) {
+    pluginOptions.prefix = (pluginOptions.prefix || '') + prefix
+  }
 
-        if (dependencies.length) {
-          cyclicDependencyCheck[name] = true
-          dependencies.forEach((name) => allPlugins[name] && loadPlugin(allPlugins[name]))
-        }
+  return {
+    plugin,
+    name: pluginMeta.name || file,
+    dependencies: pluginMeta.dependencies,
+    options: pluginOptions,
+    registered: false
+  }
+}
 
-        registerPlugin(name, plugin, options)
-      }
+function registerPlugin (fastify, meta, allPlugins, parentPlugins = {}) {
+  const { plugin, name, options, dependencies = [] } = meta
 
-      const pluginKeys = Object.keys(allPlugins)
-      for (let i = 0; i < pluginKeys.length; i++) {
-        cyclicDependencyCheck = {}
+  if (parentPlugins[name]) {
+    throw new Error('Cyclic dependency')
+  }
 
-        try {
-          loadPlugin(allPlugins[pluginKeys[i]])
-        } catch (err) {
-          next(enrichError(err))
-          return
-        }
-      }
+  if (meta.registered) {
+    return
+  }
 
+  parentPlugins[name] = true
+  for (const name of dependencies) {
+    if (allPlugins[name]) {
+      registerPlugin(fastify, allPlugins[name], allPlugins, parentPlugins)
+    }
+  }
+
+  fastify.register(plugin, options)
+  meta.registered = true
+}
+
+function wrapRoutes (content) {
+  if (content &&
+    Object.prototype.toString.call(content) === '[object Object]' &&
+    Object.prototype.hasOwnProperty.call(content, 'method')) {
+    return function (fastify, opts, next) {
```

Comment:
why not use async/await here as well? wouldn't  a throw from fastify.route then propagate up?